scalableminds/webknossos
# Reduxify volumetracing

This pull request reduxifies the volumetracing code. The code in `volumetracing.js` was refactored into actions and corresponding reducers (setActiveCell, createCell, ...). The actual cell annotation is now implemented as a saga. It also generalizes the existing `save_saga` and uses it for volumetracing changes as well.
Therefore the files `volumetracing.js, statelogger.js and volumetracing_statelogger.js` are no longer needed.

Stuff you should know:
  - Volumetracing cells are created on demand, they appear on the texture but may not have a representation in the store yet. This is why setActiveCell sometimes "creates" a new cell.
  - In difference to the skeletontracing there is only one volumetracing save update action which is `updateTracing`. The actual annotation data is sent to the server via the segmentation layers push queue.
  - Although the actual cell annotation is a saga now (which holds an instance of the `VolumeLayer` class in an instance variable during the "drawing" process), the `contourList` (which is the contour of the cell that is being drawn currently) is persisted in the store - as it should be rendered. Previously the `contourList`has been part of the `VolumeLayer` class.

Test coverage so far (more will be added shortly):
- [x] volumetracing save saga tests
- [x] cell annotation saga tests
- [x] volumetracing reducers

Steps to test:
- Run tests.
- Open volumetracing and test everything:
  - moving/tracing
  - cellId that is shown in the Tracing-Menu - Volume Options (should change when another cell is selected via shift-leftclick or when a new cell is created)
  - switch modes by clicking or shortcut (`1` or `w`)
  - create cell by clicking the button or shortcut (`c`)
  - saving (is the position and activeCellId remembered after reload)
  - does it prompt you if there are unsaved changes
  - does clicking the save button work as intended

Issues:
- fixes #1752 
- fixes #1770

------
